internal/internal_workflow_client.go

```diff
@@ -1708,45 +1747,90 @@ func (w *workflowClientInterceptor) UpdateWorkflow(
+			baseUpdateHandle: baseUpdateHandle{ref: resp.GetUpdateRef()},
+		}, nil
+	case *updatepb.Outcome_Success:
+		return &completedUpdateHandle{
+			value:            newEncodedValue(v.Success, w.client.dataConverter),
+			baseUpdateHandle: baseUpdateHandle{ref: resp.GetUpdateRef()},
+		}, nil
+	}
+	return nil, fmt.Errorf("unsupported outcome type %T", resp.GetOutcome().GetValue())
+}
+
+func (w *workflowClientInterceptor) PollWorkflowUpdate(
+	ctx context.Context,
+	in *ClientPollWorkflowUpdateInput,
+) (converter.EncodedValue, error) {
+	// header, _ = headerPropagated(ctx, w.client.contextPropagators)
+	//todo header not in PollWorkflowUpdate
+
+	grpcCtx, cancel := newGRPCContext(ctx, defaultGrpcRetryParameters(ctx))
+	defer cancel()
+	wait := updatepb.WaitPolicy{
+		LifecycleStage: enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED,
+	}
+	pollReq := workflowservice.PollWorkflowExecutionUpdateRequest{
+		Namespace:  w.client.namespace,
+		UpdateRef:  in.UpdateRef,
+		WaitPolicy: &wait,
+		Identity:   w.client.identity,
+	}
+	resp, err := w.client.workflowService.PollWorkflowExecutionUpdate(grpcCtx, &pollReq)
+	if err != nil {
+		return nil, err
+	}
+	switch v := resp.GetOutcome().GetValue().(type) {
 	case *updatepb.Outcome_Failure:
-		handle.err = w.client.failureConverter.FailureToError(v.Failure)
+		return nil, w.client.failureConverter.FailureToError(v.Failure)
 	case *updatepb.Outcome_Success:
-		handle.value = newEncodedValue(v.Success, w.client.dataConverter)
+		return newEncodedValue(v.Success, w.client.dataConverter), nil
+	default:
+		return nil, fmt.Errorf("unsupported outcome type %T", v)
 	}
-	return handle, nil
 }
 
 // Required to implement ClientOutboundInterceptor
 func (*workflowClientInterceptor) mustEmbedClientOutboundInterceptorBase() {}
 
-func (uh *updateHandle) WorkflowID() string {
+func (uh *baseUpdateHandle) WorkflowID() string {
 	return uh.ref.GetWorkflowExecution().GetWorkflowId()
 }
 
-func (uh *updateHandle) RunID() string {
+func (uh *baseUpdateHandle) RunID() string {
 	return uh.ref.GetWorkflowExecution().GetRunId()
 }
 
-func (uh *updateHandle) UpdateID() string {
+func (uh *baseUpdateHandle) UpdateID() string {
 	return uh.ref.GetUpdateId()
 }
 
-func (uh *updateHandle) Get(ctx context.Context, valuePtr interface{}) error {
-	// implementation note: this is a broken implementation that assumes the
-	// update has already completed by the time Get is called. This is true for
-	// the current implementation of synchronous updates but will not be true in
-	// the future with async update.
+func (ch *completedUpdateHandle) Get(ctx context.Context, valuePtr interface{}) error {
+	if ch.err != nil || valuePtr == nil {
+		return ch.err
+	}
+	if err := ch.value.Get(valuePtr); err != nil {
+		return err
+	}
+	return nil
+}
 
-	if uh.err != nil || valuePtr == nil {
-		return uh.err
+func (luh *lazyUpdateHandle) Get(ctx context.Context, valuePtr interface{}) error {
```

Comment:
I remember discussions around being able to wait for a stage instead of completion, is that a future PR? Also, is there any way for me to just check whether completed in a non-poll/non-blocking way?

Reply:
That would be subsequent to this, yes. This is just enough to get e2e working being async and blocking on ACCEPTED

Reply:
Spoke w/Chad  - we both see TryGet as a likely near-term request. Less clear is some form of "describe" API to observe execution state. We can wait for some usecases for the latter.